packages/block-editor/src/store/actions.js

```diff
@@ -1256,12 +1256,27 @@ export function replaceInnerBlocks(
  *
  * @return {Object} Action object.
  */
-export function toggleBlockMode( clientId ) {
-	return {
-		type: 'TOGGLE_BLOCK_MODE',
-		clientId,
+export const toggleBlockMode =
+	( clientId ) =>
+	( { select, dispatch } ) => {
+		if ( ! clientId ) {
+			return;
+		}
+		const block = select.getBlocksByClientId( clientId )[ 0 ];
+		if ( ! hasBlockSupport( block.name, 'html', true ) ) {
+			return;
+		}
+		const isCodeEditingEnabled =
+			select.getSettings().codeEditingEnabled ?? true;
+		if ( ! isCodeEditingEnabled ) {
+			return;
+		}
```

Comment:
Just noting that this logic will also prevent a block that doesn't support html etc. from being switched to visual mode. That is probably fine since assumedly the block supports and editor settings won't change after the app is initialised.